### ISSUE TITLE
Is there a way to add default abbreviation or open luapad with common local variables?

### ISSUE THREAD

**Issue by kkharji** [rafcamlet/nvim-luapad#6]
Hi @rafcamlet , hope your having an awesome weekend. 

I noticed that I'm heavily some abbreviation like `vim.fn = vfn or p = require'pkg'`, and when I want to test something from my code into luapad I often do `:%s/vfn/vim.fn/ge<cr>`, which is a bit of a drudgery, given the fact that I close and open luapad a lot.

Is there a way to add default abbreviation, or make luapad launch with a set of local abbreviation or require statement.

Thanks

**Comment by rafcamlet** [comment]
Hi!
Luapad is no different in this case from other buffers, so all defined global abbreviations should work in it too. It has double filetype "lua.luapad", so it inherits settings from lua and can have its own settings defined in ftplugin/luapad.vim file or via autocommand: For example `autocmd FileType lua.luapad iab <buffer> vfn vim.fn`

**Comment by kkharji** [comment]
Great that solves part of my problem, but it wasn't what I meant by 'add default abbreviation or open luapad with common local variables', sorry I'm sure the word abbreviation cause the missunderstanding.

Let me rephrase, I'd like to load luapad with a preset group of variables and modules.

**Comment by rafcamlet** [comment]
Oh... I get it now. Currently, there is not such option. I will add possibility to provide custom context function (evaluation environment) after next major release.

**Comment by rafcamlet** [comment]
Hi! It will take me a while to finish major update, so I've pushed some things to dev branch. It's bugged, unstable and can change without warning, so keep that in mind.

https://github.com/rafcamlet/nvim-luapad/tree/dev

There is a lua config function, which handles currently available setting and also two new things: `on_init` callback and `context` object.

```lua
require 'luapad'.config{
  on_init = function()
    vim.wo.cursorline = true
    print('Hello form Luapad!')
  end,
  context = {
    wow = 'wow',
    hello_world = function()
      print 'Hello world!' -- It is a native print
    end,
    echo = function(str)
      return str, str
    end
  }
}
```

You can use on_init to set options and run arbitrary code. Context dictionary works as luapad global environment. Remember that function passed to it, weren't defined in luapad context, so they don't use overwritten luapad print.

Have fun!

**Comment by kkharji** [comment]
Finally, I got to testing this new feature. First of all, great implementation, I love the idea of defining contexts and actually initialize hooks like  I would with autocmd but better.

Thanks a lot @rafcamlet I feel so much better that I can preset stuff I will use all the time and whenever needed composing something I'd open Luapad.

Note: I don't know maybe this got nothing to do with Luapad, but completion from nvim_lsp doesn't work when I'm in luapad buffer. I tried adding this but didn't work:

```
  require'nvim_lsp'.sumneko_lua.setup({
    on_attach = on_attach,
    filetypes = {'lua', 'lua.luapad'},
})
```
Did you face similar issue in the past. Thanks

**Comment by rafcamlet** [comment]
@tami5 I've checked nvim-lspconfig and everything is ok there. The luapad buffer is correctly registered. But it's look like whole native lsp client doesn't work with unsaved buffers with sumenko (however it works with others lsp-s) , which is weird, because I've already used sumenko through coc.nvim and everything was fine.

Sorry, but I don't know how to fix it. :(

**Comment by kkharji** [comment]
@rafcamlet That is so true, when I was using coc.nvim in my old setup this wasn't an issue. I'll try to do further testing on other lsp frameworks 

Thanks 🙏
